toolchest_client/tools/api.py

```diff
@@ -8,7 +8,7 @@
     STARInstance, Test, Unicycler
 
 
-def bowtie2(inputs, output_path=None, database_name="GRCh38_noalt_as", database_version="1", tool_args=""):
+def bowtie2(inputs, output_path=None, database_name="GRCh38_noalt_as", database_version="1", tool_args="", **kwargs):
```

Comment:
Is the `**kwargs` just to avoid adding `is_async` (and future new args) to the subclasses? I'm not really a fan of that from a readability perspective.

Reply:
As I see it, there are two categories of args:

1) Tool specific args that change from tool to tool, or at least handled differently in different tool cases (e.g. inputs could also be read_one + read_two, etc).

2) Things that apply to _all_ tools, and aren't interface specific.

I think all args for case 1 should be exposed at a top level (e.g. inputs), and the args for case 2 should be contained within kwargs. I would argue that `parallelize` and other similar args also fall into case 2, but we haven't really generalized that beyond STAR yet.

afaik `**kwargs` should be used when you want to pass handling of an undetermined amount of args down to a lower level. I agree that it's not usually the best use case, but I think in this case (passing in api.py -> tool specific implementation  (e.g. kraken2.py) -> tool.py) it makes it a lot easier.

Reply:
(the undetermined args being the other `is_async`-like args that we enable at the tool level after this point

Reply:
I see the two different argument types but I think this makes the ux generally worse. If I'm a new user of the client I probably have 1 or 2 tools I want to use that I may already be familiar with the tool args but not the toolchest args. So if I want to write `toolchest.kraken2(...)` in an ide and see the args list from there it'll list some args but then hit `**kwargs`and now the user needs to explore tool.py to find the full list. This is fine for developers but since the primary idea is to make it straightforward for bioinformaticians and other people who aren't pure developers this hurts that goal (and even as a developer not having to look up exact arg names is nice).

Reply:
As long as what can be added in `**kwargs` is documented eventually, I'm okay with this. There's another parameter (`skip_decompression`) that I want to add, but it's generalizable much like `is_async` is in this case.

Reply:
I do see the argument of adding it for usability + visibility that @jherr-dev suggested, though -- if that's our goal, then we should optimize for that now and refactor later if argument glut hinders readability and development speed.

Reply:
I agree with the argument for making this easier for bioinformaticians / etc, but this functionality – and most generalizable functionality that I'll imagine we're going to add – aren't going to fall into that use case.

People who are using `is_async` (and I would imagine `skip_decompression`) are much more competent.

Keep in mind that the specific user we're building for in this case was also okay building directly into our completely non-documented API.

Reply:
I do agree that we may eventually wan't to refactor to focus on ease of use via an IDE / etc / but I think that's premature optimization for UX in this case, at the cost of development speed.

I don't feel _too_ strongly about this and am okay with refactoring. Will ping Bryce to discuss live.

Reply:
Resolved live with Bryce